services/wfm/src/main/java/org/openkilda/wfm/topology/nbworker/bolts/LinkOperationsBolt.java

```diff
@@ -129,59 +129,84 @@ public void init() {
-                        link.setCost(linkPropsToSet.getCost());
-                    }
-                    if (linkPropsToSet.getMaxBandwidth() != null) {
-                        link.setMaxBandwidth(linkPropsToSet.getMaxBandwidth());
+                if (linkPropsToSet.getMaxBandwidth() != null) {
+                    linkProps.setMaxBandwidth(linkPropsToSet.getMaxBandwidth());
+                }
+                linkProps.setTimeModify(Instant.now());
+            } else {
+                linkProps = linkPropsToSet;
+                Instant timestamp = Instant.now();
+                linkProps.setTimeCreate(timestamp);
+                linkProps.setTimeModify(timestamp);
+            }
+            linkPropsRepository.createOrUpdate(linkProps);
+
+            Optional<Isl> existingIsl = islRepository.findByEndpoints(
+                    linkPropsToSet.getSrcSwitchId(), linkPropsToSet.getSrcPort(),
+                    linkPropsToSet.getDstSwitchId(), linkPropsToSet.getDstPort());
+            existingIsl.ifPresent(link -> {
+                if (linkPropsToSet.getCost() != null) {
+                    link.setCost(linkPropsToSet.getCost());
+                }
+                if (linkPropsToSet.getMaxBandwidth() != null) {
+                    long currentMaxBandwidth = link.getMaxBandwidth();
+                    long availableBandwidth = link.getAvailableBandwidth();
+                    if (linkPropsToSet.getMaxBandwidth() > currentMaxBandwidth) {
+                        availableBandwidth += linkPropsToSet.getMaxBandwidth() - currentMaxBandwidth;
+                    } else {
+                        availableBandwidth -= currentMaxBandwidth - linkPropsToSet.getMaxBandwidth();
                     }
-                    link.setTimeModify(Instant.now());
-
-                    islRepository.createOrUpdate(link);
-                });
+                    link.setMaxBandwidth(linkPropsToSet.getMaxBandwidth());
+                    link.setAvailableBandwidth(availableBandwidth);
+                }
+                link.setTimeModify(Instant.now());
 
-                return linkProps;
+                islRepository.createOrUpdate(link);
             });
 
-            return new LinkPropsResponse(request, LinkPropsMapper.INSTANCE.map(result), null);
-        } catch (Exception e) {
-            log.error("Unhandled exception in put linkprops operation.", e);
+            return linkProps;
+        });
+    }
 
-            return new LinkPropsResponse(request, null, e.getMessage());
-        }
+    private LinkProps swapLinkProps(LinkProps source) {
+        return LinkProps.builder()
+                .srcSwitchId(source.getDstSwitchId())
+                .srcPort(source.getDstPort())
+                .dstSwitchId(source.getSrcSwitchId())
+                .dstPort(source.getSrcPort())
+                .maxBandwidth(source.getMaxBandwidth())
+                .cost(source.getCost())
```

Comment:
fixed

Reply:
lines are still here

Reply:
still here

Reply:
Please use toBuilder() method to create whole copy of source, because you missed the other fields. It ruins the contract. We may expect that returned instance has other fields (like `time_create` or `time_create`) 

`return source.toBuilder()`
`                       .srcSwitchId(source.getDstSwitchId())`
`                        .srcPort(source.getDstPort())`
`                        .dstSwitchId(source.getSrcSwitchId())`
`                        .dstPort(source.getSrcPort())`
`                        .maxBandwidth(source.getMaxBandwidth())`
`                        .build();`